### ISSUE TITLE
BalanX lights can't initialize

### ISSUE THREAD

**Issue by roabyd** [monty68/uniled#12]
## Version of the custom_component
V0.2.5 BETA
Commit: d9d4c43d1337c47503d570b4ad0a76b8c4968567

## Configuration

Adding a SP611E light string

## Describe the bug

```text

I am trying to add this integration to control my SP611E light string, and the integration is showing up with the correct MAC address, however it is constantly switching between "Initializing" and "Retrying setup: SP611E: Data Update failed" (see below) 

```
![image](https://user-images.githubusercontent.com/55369460/208587747-252a6a5c-7ea5-467a-97fa-e20333782a6b.png)

## Debug log

```text

## Version of the custom_component
V0.2.5 BETA
Commit: d9d4c43d1337c47503d570b4ad0a76b8c4968567

## Configuration

Adding a SP611E light string

## Describe the bug

```text

I am trying to add this integration to control my SP611E light string, and the integration is showing up with the correct MAC address, however it is constantly switching between "Initializing" and "Retrying setup: SP611E: Data Update failed" (see below) 

```
![image](https://user-images.githubusercontent.com/55369460/208587747-252a6a5c-7ea5-467a-97fa-e20333782a6b.png)

## Debug log

```text

2022-12-20 15:24:09.487 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration tapo which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.492 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration reolink_dev which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.494 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration alexa_media which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.496 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration watchman which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.499 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration samsungtv_smart which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.502 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration variable which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.504 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration webrtc which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.507 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration rpi_gpio which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.509 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration dahua which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.510 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration balanx which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.513 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration hacs which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.515 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration localtuya which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.517 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration bureau_of_meteorology which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.518 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration garbage_collection which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.520 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration uniled which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 15:24:09.522 WARNING (SyncWorker_2) [homeassistant.loader] We found a custom integration govee which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant

2022-12-20 15:24:22.120 INFO (MainThread) [homeassistant.setup] Setting up uniled
2022-12-20 15:24:22.120 INFO (MainThread) [homeassistant.setup] Setup of domain uniled took 0.0 seconds

2022-12-20 15:24:24.252 DEBUG (MainThread) [custom_components.uniled] *** Setup UniLED BLE Device: FF:FF:A0:64:6C:D0: SP611E v1 - ({'device_class': 'ble', 'address': 'FF:FF:A0:64:6C:D0', 'model': 'SP611E'})
2022-12-20 15:24:24.252 DEBUG (MainThread) [custom_components.uniled.lib.classes] SP611E: Init Channel 0
2022-12-20 15:24:24.252 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Init BLE Device (Model: SP611E)
2022-12-20 15:24:24.252 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Set model SP611E
2022-12-20 15:24:24.252 DEBUG (MainThread) [custom_components.uniled.coordinator] SP611E: Update, entry state: ConfigEntryState.SETUP_IN_PROGRESS
2022-12-20 15:24:24.253 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Update
2022-12-20 15:24:24.253 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Connecting; RSSI: -68

2022-12-20 15:24:42.949 DEBUG (MainThread) [custom_components.uniled.coordinator] Finished fetching SP611E data in 18.696 seconds (success: False)
2022-12-20 15:24:42.949 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Stop
2022-12-20 15:24:42.949 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Disconnecting from device
2022-12-20 15:24:42.949 WARNING (MainThread) [homeassistant.config_entries] Config entry 'SP611E' for uniled integration not ready yet: SP611E: Data Update failed; Retrying in background

2022-12-20 15:24:46.152 DEBUG (Thread-4) [custom_components.uniled.coordinator] Coordinator destroyed
2022-12-20 15:24:46.152 DEBUG (Thread-4) [custom_components.uniled.lib.ble_device] Device Destroyed
2022-12-20 15:24:46.152 DEBUG (Thread-4) [custom_components.uniled.lib.classes] Channel 0 Destroyed

2022-12-20 15:25:02.256 DEBUG (MainThread) [custom_components.uniled] *** Setup UniLED BLE Device: FF:FF:A0:64:6C:D0: SP611E v1 - ({'device_class': 'ble', 'address': 'FF:FF:A0:64:6C:D0', 'model': 'SP611E'})
2022-12-20 15:25:02.256 DEBUG (MainThread) [custom_components.uniled.lib.classes] SP611E: Init Channel 0
2022-12-20 15:25:02.256 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Init BLE Device (Model: SP611E)
2022-12-20 15:25:02.256 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Set model SP611E
2022-12-20 15:25:02.257 DEBUG (MainThread) [custom_components.uniled.coordinator] SP611E: Update, entry state: ConfigEntryState.SETUP_IN_PROGRESS
2022-12-20 15:25:02.257 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Update
2022-12-20 15:25:02.280 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Connecting; RSSI: -68

2022-12-20 15:25:39.601 DEBUG (MainThread) [custom_components.uniled] *** Setup UniLED BLE Device: FF:FF:A0:64:6C:D0: SP611E v1 - ({'device_class': 'ble', 'address': 'FF:FF:A0:64:6C:D0', 'model': 'SP611E'})
2022-12-20 15:25:39.602 DEBUG (MainThread) [custom_components.uniled.lib.classes] SP611E: Init Channel 0
2022-12-20 15:25:39.602 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Init BLE Device (Model: SP611E)
2022-12-20 15:25:39.602 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Set model SP611E
2022-12-20 15:25:39.603 DEBUG (MainThread) [custom_components.uniled.coordinator] SP611E: Update, entry state: ConfigEntryState.SETUP_IN_PROGRESS
2022-12-20 15:25:39.603 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Update
2022-12-20 15:25:39.603 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Connecting; RSSI: -68

2022-12-20 15:25:40.696 DEBUG (MainThread) [custom_components.uniled.coordinator] Finished fetching SP611E data in 1.093 seconds (success: False)
2022-12-20 15:25:40.697 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Stop
2022-12-20 15:25:40.697 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Disconnecting from device

2022-12-20 15:25:46.176 INFO (MainThread) [homeassistant.components.bluetooth.scanner] hci0 (E4:5F:01:8F:85:88): Bluetooth scanner has gone quiet for 90s, restarting
2022-12-20 15:25:46.320 WARNING (MainThread) [bluetooth_auto_recovery.recover] Bluetooth adapter hci0 [E4:5F:01:8F:85:88] successfully turned back ON


2022-12-20 15:25:54.129 DEBUG (SyncWorker_4) [custom_components.uniled.coordinator] Coordinator destroyed
2022-12-20 15:25:54.130 DEBUG (SyncWorker_4) [custom_components.uniled.lib.ble_device] Device Destroyed
2022-12-20 15:25:54.130 DEBUG (SyncWorker_4) [custom_components.uniled.lib.classes] Channel 0 Destroyed

2022-12-20 15:26:20.699 DEBUG (MainThread) [custom_components.uniled] *** Setup UniLED BLE Device: FF:FF:A0:64:6C:D0: SP611E v1 - ({'device_class': 'ble', 'address': 'FF:FF:A0:64:6C:D0', 'model': 'SP611E'})
2022-12-20 15:26:20.700 DEBUG (MainThread) [custom_components.uniled.lib.classes] SP611E: Init Channel 0
2022-12-20 15:26:20.700 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Init BLE Device (Model: SP611E)
2022-12-20 15:26:20.700 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Set model SP611E
2022-12-20 15:26:20.701 DEBUG (MainThread) [custom_components.uniled.coordinator] SP611E: Update, entry state: ConfigEntryState.SETUP_IN_PROGRESS
2022-12-20 15:26:20.701 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Update
2022-12-20 15:26:20.702 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Connecting; RSSI: -68
2022-12-20 15:26:22.371 ERROR (MainThread) [homeassistant.components.alexa.state_report] Timeout sending report to Alexa for sensor.processor_temperature

```

```

**Comment by monty68** [comment]
Please try again with v0.2.7 Beta, this may not resolve the issue but will provide me with some extra debugging information.

I note in your debug output a custom component named "balanx" is this also a component to control BLE devices such as the SP611E? If so can you ensure this is disabled before retrying?

**Comment by roabyd** [comment]
I have just updated and now seeing these logs:

```text
2022-12-20 19:09:53.207 WARNING (SyncWorker_1) [homeassistant.loader] We found a custom integration uniled which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 19:09:53.208 WARNING (SyncWorker_1) [homeassistant.loader] We found a custom integration govee which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-12-20 19:10:07.174 DEBUG (MainThread) [custom_components.uniled] *** Setup UniLED BLE Device: FF:FF:A0:64:6C:D0: SP611E v1 - ({'device_class': 'ble', 'address': 'FF:FF:A0:64:6C:D0', 'model': 'SP611E'}), Service: <home_assistant_bluetooth.models.BluetoothServiceInfoBleak object at 0x7f75256b40>
2022-12-20 19:10:07.174 DEBUG (MainThread) [custom_components.uniled.lib.classes] SP611E: Init Channel 0
2022-12-20 19:10:07.174 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Init BLE Device (Model: SP611E)
2022-12-20 19:10:07.174 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Set model SP611E
2022-12-20 19:10:07.175 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Update
2022-12-20 19:10:07.175 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Connecting; RSSI: -68

2022-12-20 19:10:27.292 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: connection failed: SP611E - FF:FF:A0:64:6C:D0: Failed to connect: [org.bluez.Error.Failed] Input/output error; RSSI: -68
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/bleak_retry_connector/__init__.py", line 625, in establish_connection
    await client.connect(
  File "/usr/src/homeassistant/homeassistant/components/bluetooth/wrappers.py", line 203, in connect
    connected = await super().connect(**kwargs)
  File "/usr/local/lib/python3.10/site-packages/bleak/__init__.py", line 471, in connect
    return await self._backend.connect(**kwargs)
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/client.py", line 190, in connect
    assert_reply(reply)
  File "/usr/local/lib/python3.10/site-packages/bleak/backends/bluezdbus/utils.py", line 20, in assert_reply
    raise BleakDBusError(reply.error_name, reply.body)
bleak.exc.BleakDBusError: [org.bluez.Error.Failed] Input/output error

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/config/custom_components/uniled/lib/ble_device.py", line 213, in send_command
    await self._ensure_connected()
  File "/config/custom_components/uniled/lib/ble_device.py", line 248, in _ensure_connected
    client = await establish_connection(
  File "/usr/local/lib/python3.10/site-packages/bleak_retry_connector/__init__.py", line 702, in establish_connection
    _raise_if_needed(name, device.address, exc)
  File "/usr/local/lib/python3.10/site-packages/bleak_retry_connector/__init__.py", line 580, in _raise_if_needed
    raise BleakConnectionError(msg) from exc
bleak_retry_connector.BleakConnectionError: SP611E - FF:FF:A0:64:6C:D0: Failed to connect: [org.bluez.Error.Failed] Input/output error
2022-12-20 19:10:27.297 DEBUG (MainThread) [custom_components.uniled.coordinator] Finished fetching SP611E data in 20.122 seconds (success: False)
2022-12-20 19:10:27.297 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Stop
2022-12-20 19:10:27.297 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] SP611E: Disconnecting from device
2022-12-20 19:10:27.298 WARNING (MainThread) [homeassistant.config_entries] Config entry 'SP611E' for uniled integration not ready yet: SP611E: Data Update failed; Retrying in background
2022-12-20 19:10:28.633 DEBUG (MainThread) [custom_components.uniled.coordinator] Coordinator destroyed
2022-12-20 19:10:28.634 DEBUG (MainThread) [custom_components.uniled.lib.ble_device] Device Destroyed
2022-12-20 19:10:28.634 DEBUG (MainThread) [custom_components.uniled.lib.classes] Channel 0 Destroyed
```

Hopefully this helps

**Comment by monty68** [comment]
Thank you for the logs, but this suggests this is a communication error with the device rather than an issue with the UniLED integration itself. I don't know enough about the lower-level bleak library to help any further.

**Comment by roabyd** [comment]
Confirmed this was a physical issue with the build in Bluetooth adapter on the RPI 4B, adding a comment for anyone else facing a similar issue. 

I had recently bought a metal case for the Pi and it looks like this caused too much interference with the onboard Bluetooth adapter. I have now purchased an external BLE adapter (Kinivo BTD-400 found in the docs) with a USB extension cable and this is now working like a charm!

Thanks so much for your great work and help investigating.
